torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2295,6 +2352,64 @@ def _rebuild_full_params(self) -> List[Tuple[torch.Tensor, bool]]:
                         self._free_mp_shard(cast(List[FlatParameter], [p]))
         return output_tensors
 
+    def _check_all_gather(self, param: FlatParameter):
+        """
+        Checks the validity of an all-gather to rebuild the full parameter
+        ``param``. If on the first iteration, this uses an all-gather to check
+        that all ranks plan to all-gather the same parameter, erroring if not,
+        and on subsequent iterations, this warns the user if the all-gather
+        execution order differs from that of the first iteration, meaning that
+        we can no longer guarantee correct execution.
+
+        Executing in ``no_sync()`` does not affect the execution order check
+        for ``FULL_SHARD`` and ``SHARD_GRAD_OP``: (1) Being in ``no_sync()``
+        in the first iteration does not yield a different all-gather sequence,
+        and (2) being in ``no_sync()`` in a later iteration does not give
+        false positive warnings since the all-gather sequence still matches the
+        first iteration sequence (for ``FULL_SHARD``) or the first iteration
+        sequence's prefix (for ``SHARD_GRAD_OP``).
+        """
+        # Only check all-gathers when rebuilding the full parameters in the
+        # forward pass or in the beginning of the backward pass
+        if self.training_state not in (
+            TrainingState_.FORWARD, TrainingState_.BACKWARD_PRE,
+        ):
+            return
+        eod = self._exec_order_data
+        param_index = eod.param_to_param_index[param]
+        if not eod.is_first_iter:
+            # Warn possibly multiple times for the same iteration since the
+            # number of warnings may be informative in debugging
+            if eod.index >= len(eod.param_order) and self.rank == 0:
+                print("eod.index:", eod.index)
+                print("eod.param_order:", eod.param_order)
+            if param_index != eod.param_order[eod.index]:
+                warnings.warn(
+                    "Execution order differs from that of the first iteration "
+                    f"on rank {self.rank}; collectives are unchecked and may "
+                    "give incorrect results or hang\n"
+                    f"First iter: {eod.param_order}\nThis iter (so far): "
+                    f"{eod.param_order[:eod.index - 1] + [param_index]}"
+                )
+            eod.index += 1
+        else:
+            device = param.device
+            indices = torch.zeros(self.world_size, dtype=torch.int32).to(device)
+            index = torch.tensor([param_index], dtype=torch.int32).to(device)
+            dist._all_gather_base(indices, index, group=self.process_group)
```

Comment:
what happens if user is gathering the same parameter index but the module somehow has different shape? Does broadcast at the beginning or torch_distributed_debug anticipated to help with this?

Reply:
I agree then it is problematic. I guess with your diff for checking parameter shapes, this is not an issue since the shapes will already be validated.